File: haskell/glue/src/Glue/Lib/Math/Trigonometric/Tan.hs
module Glue.Lib.Math.Trigonometric.Tan where

import Glue.Eval (Eval, throwError)
import Glue.Eval.Exception (wrongArgumentType)
import Glue.IR (IR (..))

-- Tangent function
-- Mirrors Haskell Glue.Lib.Math.Trigonometric.Tan.tan exactly
tan :: IR Eval
tan = NativeFunc tanImpl

-- Tangent function implementation
-- Mirrors Haskell Glue.Lib.Math.Trigonometric.Tan.tanImpl exactly
tanImpl :: IR Eval -> Eval (IR Eval)
tanImpl arg = case arg of
    Integer n -> pure $ Float (Prelude.tan (fromIntegral n))
    Float n -> pure $ Float (Prelude.tan n)
    _ -> throwError $ wrongArgumentType ["number"]

File: dart/glue/lib/src/lib/math/trigonometric/tan.dart
import 'dart:math' as math;

import 'package:glue/src/eval.dart';
import 'package:glue/src/eval/exception.dart';
import 'package:glue/src/ir.dart';

/// Tangent function (radians)
/// Mirrors Haskell Glue.Lib.Math.Trigonometric.Tan.tan exactly
final Ir tan = IrNativeFunc(tanImpl);

/// Tangent function implementation (radians)
/// Mirrors Haskell Glue.Lib.Math.Trigonometric.Tan.tanImpl exactly
Eval<Ir> tanImpl(Ir arg) {
  return eval(arg).flatMap((va) {
    return switch (va) {
      IrInteger(value: final n) => Eval.pure(IrFloat(math.tan(n.toDouble()))),
      IrFloat(value: final n) => Eval.pure(IrFloat(math.tan(n))),
      _ => throwError(wrongArgumentType(['number'])),
    };
  });
}
